### PR TITLE
use getopt; improve help output; glibc < 2.28 compatibility

### PR DESCRIPTION
**Pull request by mikegogulski**
Hello. Thanks for writing this!

### CODE REVIEW

File: exchange.c [hunk 1]
```diff
@@ -1,51 +1,76 @@
...
 
-int main(int argc, char** argv) {
-    if (argc < 3) {
-        printHelp();
-        return 1;
-    }
+void usage(int exit_code) {
+    puts("Usage: exchange [-v] PATH1 PATH2\n"
+         "  or:  exchange -h|--help\n"
+         "atomically exchange names and contents of two files or directories\n"
```
Comment (AbhyudayaSharma) [inline exchange.c]:
```suggestion
         "   or:  exchange -h|--help\n"
         "atomically exchange names of two files or directories\n"
```

According to `man renameat2`, the contents are not modified. Just the name is swapped.

File: exchange.c [hunk 2]
```diff
@@ -1,51 +1,76 @@
...
-        return 1;
+        exit(1);
     }
 
     if (verbose) {
-        puts("Names exchanged successfully");
+        puts("Paths exchanged successfully");
     }
-
-    return 0;
```
Comment (AbhyudayaSharma) [inline exchange.c:50]: I would suggest keeping the `return 0` here.

File: README.md [hunk 3]
```diff
@@ -22,9 +22,10 @@ make install
...
-exchange: exchange names of two files or directories atomically
-Usage: exchange FILE1 FILE2 [flags...]
-Allowed flags:
-        --help:    shows this help text
-        --verbose: produces verbose output
+exchange: atomically exchange names and contents of two files or directories
+Usage: exchange [OPTIONS] PATH1 PATH2
+
+Options:
+        --help     show help text and exit
```
Comment (AbhyudayaSharma) [inline README.md]: I think we should document the short options here as well.

File: exchange.c [hunk 4]
```diff
@@ -1,51 +1,76 @@
...
 
-    // atomically exchange the two directories
-    // if paths are not absolute, consider them relative to the current working directory
-    if(renameat2(AT_FDCWD, argv[1], AT_FDCWD, argv[2], RENAME_EXCHANGE) == -1) {
+    // Atomically exchange the two paths. If paths are not absolute, consider
+    // them relative to the current working directory
+#if (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 28))
+        ret = renameat2(AT_FDCWD, paths[0], AT_FDCWD, paths[1], RENAME_EXCHANGE);
+#else
+        ret = syscall(SYS_renameat2, AT_FDCWD, paths[0], AT_FDCWD, paths[1], RENAME_EXCHANGE);
```
Comment (AbhyudayaSharma) [inline exchange.c:65]: Does the `RENAME_EXCHANGE` constant exist in glibc < 2.28? I don't have immediate access to a system with an old glibc version. We should probably test this in Docker.

**Review by AbhyudayaSharma — COMMENTED** [review]
Hi @mikegogulski ! Thanks a lot for the pull request. Looks good to me apart from some minor concerns.

**Comment by AbhyudayaSharma** [comment]
Somehow GitHub actions were not triggered for this PR. I'll need to investigate that.

**Comment by AbhyudayaSharma** [comment]
Closing and reopening to trigger the CI workflow.

File: exchange.c (hunk 1 above)
Comment (mikegogulski) [reply exchange.c]: Fair enough. I find it a bit easier to reason about "names and contents" for what renameat2 does, but it's a very narrow point. I'll remove that.

File: exchange.c (hunk 2 above)
Comment (mikegogulski) [reply exchange.c:50]: `return 0` is implicit, but that's fine. I went a step further and replaced 0 and 1 exit codes with EXIT_SUCCESS and EXIT_FAILURE from stdlib.h.

File: README.md (hunk 3 above)
Comment (mikegogulski) [reply README.md]: Fixed.

File: exchange.c (hunk 4 above)
Comment (mikegogulski) [reply exchange.c:65]: It does. I was led to add this workaround when, after finding everything worked nicely on Ubuntu 20.04, I went to try to build it on the 18.04 system where I actually need it, which has glibc 2.27. `RENAME_EXCHANGE` works with the `syscall` there, but the `renameat2` function is missing.

I'll push an update commit in a sec.

Comment (mikegogulski) [reply exchange.c:65]: If we were really clever we'd also test for the kernel version below which SYS_renameat2 didn't exist.

Citation for some of this stuff:

"renameat2() was added to Linux in kernel 3.15; library support was added in glibc 2.28."

https://man7.org/linux/man-pages/man2/rename.2.html

**Review by AbhyudayaSharma — APPROVED** [review]
Thanks a lot @mikegogulski !

**Pull request merged**